alphagov/paas-cf
# [#134175741] create-user.sh: Replace passwords with invites

## What

Use UAA's builtin mechanism for generating (but not sending) invites for new
accounts. A link with a one-time use code allows the user to set their own
password. The code is invalidated after first use. This is preferable to
generating passwords because we don't need to send them in plaintext email
and we don't need to recommend that people change their password after the
first login.

## How to review

Code:

1. Read the individual commits in order for more details about the implementation.
1. Forgive me for shell scripting atrocities.

Testing:

1. Deploy the branch to your dev environment
1. Run `./scripts/create_user.sh` to create yourself an account in your dev environment
1. Follow the link in the email that you receive
1. Confirm that you see a link to the documentation after setting your password
1. Confirm that you can login with the CLI using your new account
1. Run `./scripts/create_user.sh` with the `-r` option to send a new invite
1. Confirm that you can no longer login with the CLI
1. Confirm that the new invite works
1. Review alphagov/paas-tech-docs#32 also

Optional:

1. Confirm that it works with an SSO account in dev (I've done this once, but not recently)
1. Confirm that normal users don't have tokens with `scim.invite` scope, you can use:

       cf oauth-token | awk -F. '{print $2}' | (base64 -D; echo -n '}') | jq

## Who can review

@jollery or @tomskerous may want to review the email content.
